pkg/gatewayserver/io/udp/firewall_ratelimit.go

```diff
@@ -0,0 +1,93 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package udp
+
+import (
+	"sync"
+	"time"
+
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	encoding "go.thethings.network/lorawan-stack/pkg/ttnpb/udp"
+)
+
+var (
+	// defaultMessages is the default number of messages to use for rate limiting.
+	defaultMessages = 10
+
+	// defaultThreshold is the duration after which a new packet will be accepted.
+	defaultThreshold = 10 * time.Millisecond
+)
+
+type rateLimitingFirewall struct {
+	f Firewall
+
+	m sync.Map // eui to incoming
+
+	messages  int
+	threshold time.Duration
+}
+
+// NewRateLimitingFirewall returns a Firewall with rate limiting capabilities.
+func NewRateLimitingFirewall(firewall Firewall, messages int, threshold time.Duration) Firewall {
+	if messages == 0 {
+		messages = defaultMessages
+	}
+	if threshold == 0 {
+		threshold = defaultThreshold
+	}
+
+	f := &rateLimitingFirewall{
+		f:         firewall,
+		messages:  messages,
+		threshold: threshold,
+	}
+	return f
+}
+
+var (
+	errRateExceeded = errors.DefineResourceExhausted("rate_exceeded", "gateway traffic exceeded allowed rate")
+)
+
+func (f *rateLimitingFirewall) Filter(packet encoding.Packet) error {
+	if packet.GatewayEUI == nil {
+		return errNoEUI
+	}
+	if packet.GatewayAddr == nil {
+		return errNoAddress
+	}
+	now := time.Now().UTC()
+	eui := *packet.GatewayEUI
+	val, ok := f.m.Load(eui)
+	var timestamps *incoming
+	if ok {
+		timestamps = val.(*incoming)
+		oldestTimestamp := timestamps.Oldest()
+		if !oldestTimestamp.IsZero() && now.Sub(oldestTimestamp) < f.threshold {
```

Comment:
The `IsZero()` check is probably not needed, since if that is true, then the second condition will most likely fail